0035-ilp-over-http/0035-ilp-over-http.md

```diff
@@ -56,3 +52,119 @@ All ILP Packets MUST be returned with the HTTP status code `200: OK`.
+
+### Follow Standardized Security Recommendations
+It is advisable to follow all applicable best practices when using a Bearer-token scheme for authentication. [Section 5.1 of RFC-6570](https://tools.ietf.org/html/rfc6750#section-5) contains a number of very good practices that should be considered on a per-deployment basis. 
+
+### HTTP BASIC/FORM Auth
+HTTP Auth schemes using a username and password is NOT RECOMMENDED for the same reasons that the `SIMPLE` profile is only recommended for development and testing scenarios only.
+ 
+### Use SIMPLE Profile for Development/Testing Only
+The `SIMPLE` authentication profile provides only marginal benefits when compared to the `JWT_HS_256` profile, but introduces significant drawbacks as outlined in the "Trade-off Summary" sections of this RFC. As such, the `SIMPLE` profile MAY be used for development or testing purposes, but SHOULD NOT be used in production scenarios. Instead, prefer `JWT_HS_256` for production deployments.
+
+### Use Reasonable Token Expiries
+Tokens generators should choose a reasonable token expiry. Considerations in this choice include the ability to cache and re-use authentication tokens for a limited time in order to enable very fast authentication decisions. However, this should be balanced against a desire for shorter token lifetimes, which will limit the attack surface caused by a compromised token. 
+
+As a best practice, implementations SHOULD use tokens that expire. For example, consider generating tokens with a lifetime that doesn't exceed 5 minutes.
+
+### Secrets At Rest
+In most ILP Connector implementations, secrets are stored on a per-account basis in some sort of data-store. Implementations SHOULD protect secret-values that can be used to generate authentication tokens by encrypting them prior to storage. This will help prevent actual shared secrets from being captured by unauthorized parties, increasing the chances that only the Connector runtime will be able to generate tokens using underlying shared-secrets.
+
+One option is to encrypt any secrets at-rest using an Authenticated Encryption algorithm, such as AES-GCM-256. This [article](https://proandroiddev.com/security-best-practices-symmetric-encryption-with-aes-in-java-7616beaaade9) provides a nice overview using Java, along with various configuration options to consider.
+  
+### Secrets In Memory
+Implementations SHOULD minimize the amount of time that an actual secret-value exists in-memory in unencrypted form. This includes narrowing the availability of secrets to only code that actually requires them; minimizing the time any secret might exist in memory; and zeroing out memory after a secret is no longer used, if possible. 
+
+### Mutual TLS
+All ILP-over-HTTP connections MUST be performed over a TLS session. However, it is also RECOMMENDED to use TLS Client Certificates between peers for additional security. 
```

Comment:
Mutual TLS authentication is the industry standard for online payment APIs, particularly between things like banks and financial services companies.

If you're using mutual TLS, you've already established the peer identity, and there's no need to use a secondary mechanism (e.g. JWT claims) to do as such.

Unlike bearer credentials such as JWTs, X.509 peer identity is bound to the secure channel. This has pros and cons (mutual TLS makes authentication through a proxy difficult, where tokens generally carry higher exfiltration risk, and JWTs in particular risk immature/deficient implementations of a standard with a poorly thought out, overcomplicated cryptographic design), but I think the most important is that determining the validity of the remote peer can be handled directly by the TLS stack / application framework, as opposed to writing some ad hoc introspection logic for JWTs.

I think X.509 + token based solutions can be helpful, but much less so when the token is claims-bearing (which in my book means an AuthN credential). A much better use of tokens-over-mutual-TLS, IMO, is using tokens for fine-grained AuthZ and floating AuthZ policies. If you'd like to explore this, my recommendation would be to start by keeping the wire format of the tokens opaque, but stipulating a standard authorization workflow (though I dislike it for several reasons, OAuth2 fits the bill).

tl;dr: I would like to see this section expanded to include ways of obtaining identity information from the X.509 certificate presented by the peer during the TLS handshake. In particular I think it'd be interesting to explore ways of using Subject Alternative Names (SANs) in conjunction with things like ILP addresses.

Reply:
Hey @tarcieri, thanks for your input here!

> If you're using mutual TLS, you've already established the peer identity, and there's no need to use a secondary mechanism (e.g. JWT claims) to do as such.

In theory, this is true, and two years ago I would have agreed with you -- but in practice I've found that "there's no need to use a secondary mechanism" to be incorrect. The primary reason is that _many_ entities (you can use "banks" here, but it's broader than just banks) terminate TLS connections at the network edge, and then proxy those connections (sometimes over HTTP, believe it or not) to the actual backend.

In these deployments (where TLS is terminated), the statement that "X.509 peer identity is bound to the secure channel" is no longer true. The concept of authenticated identity has now been shifted to be: "Whoever the edge device says you are", which IMHO is a problem. 

> I think X.509 + token based solutions can be helpful

In my experience, because of the way entities outside of the control of software developers deploy their TLS connections, there _must_ be a higher-level authentication scheme that exists outside of TLS.

> If you'd like to explore this, my recommendation would be to start by keeping the wire format of the tokens opaque, but stipulating a standard authorization workflow

Setting aside authorization for a moment, I think your suggestion about "keeping the wire format of the tokens opaque" is the opinion I held about a month ago (i.e., keeping the tokens opaque), but after trying to build something, I ran into the trade-offs outlined in the PR (see [here](https://github.com/interledger/rfcs/pull/531/files#diff-ccfda9b7df50a9cfd8618bb9c08e360aR83)) when using a `SIMPLE` token. To your point, this is because that scheme lives "on top of" any Mutual TLS authentication that might exist.

I'm open to more feedback/ideas here though. The only reason I arrived at the `JWT_HS_256` profile defined in this PR is because of the constraints I outlined (i.e., can't rely upon TLS sessions not getting terminated + simple tokens have drawbacks).

> In particular I think it'd be interesting to explore ways of using Subject Alternative Names (SANs) in conjunction with things like ILP addresses.

I've explored similar ideas for other payment networks, and think this could be a very interesting option for a part of an ILP identity layer. The questions are many though - I've started a [forum thread here to discuss](https://forum.interledger.org/t/are-ilp-addresses-tied-to-identity-at-all-are-they-ephemeral/605). Would love to hear more thoughts from you and others on that one.

Reply:
> In these deployments (where TLS is terminated), the statement that "X.509 peer identity is bound to the secure channel" is no longer true. The concept of authenticated identity has now been shifted to be: "Whoever the edge device says you are", which IMHO is a problem.

I noted this to a certain extent earlier when I enumerated pros and cons. However, I'll also note that with a token-based solution, compromise of an edge terminator has an analogous but slightly less severe problem in that attackers can replay observed tokens (unless they are sufficiently time limited).

This is the fundamental risk of edge termination. I'll quickly note an alternative to edge termination is SNI inspection.

All that said: the information being provided is peer identity, a.k.a. AuthN/"claims". To avoid compromise of the edge terminator becoming a catastrophe, as I mentioned earlier supplemental AuthZ credentials can also be required.

The advantage of relying on mutual TLS for identity and using tokens for AuthZ only is the token can be channel bound / "contextually confined" to a particular secure channel. This won't stop an attacker who pops the edge terminator, but it will stop one who exfiltrates a client-side token.

As written this suggests using mutual TLS, then ignoring the TLS peer information, in which case why are you using mutual TLS? Making the peer identity information encoded in the token a caveat/assertion (as in an AuthZ based scheme), rather than making the token the authority for it (as you're proposing), allows you to actually leverage the security provided by the MTLS channel.

The main disadvantage of relying on bearer credentials alone is exfiltration risk: there are many secure storage options for TLS private keys, including TPMs/TEEs/SEPs as well as HSMs and hardware tokens. It's significantly easier to keep them out-of-band of application logic and thereby prevent unintentional exposure.

I'll quickly note there's [RFC8471 Token Binding](https://tools.ietf.org/html/rfc8471) as a general purpose solution for binding tokens to a secure channel using an automatically provisioned self-signed certificate, however it's quite complex and not widely implemented (I believe Chrome has abandoned their efforts to implement it).

I'll footnote this by saying you can find me arguing against client certificates in an [unrelated conversation about API authentication for Rust's crates.io package repository](https://internals.rust-lang.org/t/ultra-pre-rfc-client-certificates-for-cargo-instead-of-shared-tokens/10173/8?u=bascule). The difference there is the tokens are being used to identify human principals rather than services.

But as I noted in that thread as well: "mutual TLS vs tokens" isn't a dichotomy. The two can be combined and leveraged as a force multiplier. I'd suggest you don't paint yourself into a corner and do attempt to leverage TLS peer identity info when available.